### PR TITLE
[WIP] Create PHI nodes for Numba IR

### PR DESCRIPTION
**Pull request by guilhermeleobas**
Following PR #4149. New changes starts at [a898fb4](https://github.com/numba/numba/commit/a898fb4d7205f4a4f71b95bc39056e67fb27a7da)

This is still a working in program and there is still a lot of things to do.

### CODE REVIEW

**Comment by guilhermeleobas** [comment]
Sorry for the delay on this PR.

A quick update on what has been done so far:

- [x] Dominator Trees
- [x] `PHI` class the inherit from `Stmt`
- [x] Algorithm for placing PHI nodes
- [ ] Renaming variables. 

I'd like some feedback on what has been done so far, especially because I am not so sure of some changes.

File: numba/compiler.py [hunk 1]
```diff
@@ -816,6 +824,7 @@ def define_nopython_pipeline(self, pm, name='nopython'):
         pm.create_pipeline(name)
         self.add_preprocessing_stage(pm)
         self.add_with_handling_stage(pm)
+        self.add_ssa_transformation_stage(pm)
```
Comment (guilhermeleobas) [inline numba/compiler.py]: Is this the correct way to introduce the SSA pass?

File: numba/ir.py [hunk 2]
```diff
@@ -691,6 +691,33 @@ def __str__(self):
         return 'print(%s)' % ', '.join(str(v) for v in self.args)
 
 
+class PHI(Stmt):
```
Comment (guilhermeleobas) [inline numba/ir.py]: Should PHI inherit from `Assign`?

**Comment by DrTodd13** [comment]
@guilhermeleobas How are you testing this right now?  The full suite in runtests?

**Comment by guilhermeleobas** [comment]
> @guilhermeleobas How are you testing this right now? The full suite in runtests?

No. I am testing on a couple of python files I wrote myself and can analyze the IR by hand.

**Comment by DrTodd13** [comment]
> > @guilhermeleobas How are you testing this right now? The full suite in runtests?
> 
> No. I am testing on a couple of python files I wrote myself and can analyze the IR by hand.

I guess the answer to your question depends on how similarly assignments and phis need to be treated in two contexts: parallel=True and parallel=False.  Probably a lot of the time, they will be treated equivalently so if you do derive phi from assign then all the isinstance(..., Ir.Assign) calls will still fire and it may be something reasonable but then you have to identify and add some code for the cases where you want to do something different for phi nodes.  Unfortunately, if you don't derive from assign then it will probably silently skip phi nodes.  I guess I am leaning towards having it derive from assign but understanding the ramifications of this change throughout the whole pipeline is going to be time consuming.  I wish we had this when we started.  It would have made things easier.  But now we will need to go back and understand where we can simplify our analyses under the assumption of SSA.

**Comment by ehsantn** [comment]
Why can't `phi` be an op of `Expr` node, which is right-hand side of assignment, similar to other operators? for example, `a = PHI(a1, a2)`.

**Pull request closed without merging**